### PR TITLE
admission: add validation plugin for cluster resource quota

### PR DESCRIPTION
**Pull request by mfojtik**
Adding admission validation for cluster resource quota.
This will prepare us to switch cluster resource quota to CRD.

/cc @deads2k 
/cc @derekwaynecarr

### CODE REVIEW

File: pkg/admission/customresourcevalidation/clusterresourcequota/validation/validation.go [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+	}
+	internalStatus := &core.ResourceQuotaStatus{}
+	if err := v1.Convert_v1_ResourceQuotaStatus_To_core_ResourceQuotaStatus(&clusterquota.Status.Total, internalStatus, nil); err != nil {
+		panic(err)
+	}
+
+	allErrs = append(allErrs, validation.ValidateResourceQuotaSpec(internalQuota, field.NewPath("spec", "quota"))...)
+	allErrs = append(allErrs, validation.ValidateResourceQuotaStatus(internalStatus, field.NewPath("status", "overall"))...)
+
+	orderedNamespaces := clusterquota.Status.Namespaces.DeepCopy()
```
Comment (mfojtik) [inline pkg/admission/customresourcevalidation/clusterresourcequota/validation/validation.go]: @derekwaynecarr @deads2k this need to be reviewed as it is not using the ordered map hack we have on internal type.

File: pkg/admission/customresourcevalidation/clusterresourcequota/validate_crq.go [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+	if !ok {
+		return nil, append(allErrs,
+			field.NotSupported(field.NewPath("kind"), fmt.Sprintf("%T", uncastObj), []string{"ClusterResourceQuota"}),
+			field.NotSupported(field.NewPath("apiVersion"), fmt.Sprintf("%T", uncastObj), []string{"quota.openshift.io/v1"}))
+	}
+
+	return obj, nil
+}
+
+func toAppliedClusterResourceQuota(uncastObj runtime.Object) (*quotav1.AppliedClusterResourceQuota, field.ErrorList) {
```
Comment (deads2k) [inline pkg/admission/customresourcevalidation/clusterresourcequota/validate_crq.go]: no.  appliedclusterresourcequota is a virtual resource, so it must stay aggregated.

File: pkg/admission/customresourcevalidation/clusterresourcequota/validate_crq.go [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+	// TODO: We should probably validate the real status here?
+
+	errs = append(errs, validation.ValidateObjectMetaUpdate(&appliedClusterResourceQuotaObj.ObjectMeta, &appliedClusterResourceQuotaOldObj.ObjectMeta, field.NewPath("metadata"))...)
+	return errs
+}
+
+type clusterResourceQuotaV1 struct {
+}
+
+func (clusterResourceQuotaV1) ValidateCreate(obj runtime.Object) field.ErrorList {
```
Comment (deads2k) [inline pkg/admission/customresourcevalidation/clusterresourcequota/validate_crq.go]: where is this called from?

Comment (deads2k) [reply pkg/admission/customresourcevalidation/clusterresourcequota/validate_crq.go]: > where is this called from?

oh neat, reusing the one I build for config.openshift.io.  Good idea.

File: pkg/admission/customresourcevalidation/clusterresourcequota/validate_crq.go [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+	quotavalidation "github.com/openshift/origin/pkg/admission/customresourcevalidation/clusterresourcequota/validation"
+)
+
+const PluginName = "quota.openshift.io/ValidateClusterResourceQuota"
+
+func Register(plugins *admission.Plugins) {
+	plugins.Register(PluginName, func(config io.Reader) (admission.Interface, error) {
+		return customresourcevalidation.NewValidator(
+			map[schema.GroupResource]bool{
+				{Group: quotav1.GroupName, Resource: "clusterresourcequota"}:        true,
```
Comment (deads2k) [inline pkg/admission/customresourcevalidation/clusterresourcequota/validate_crq.go]: resources are lowercase plural

File: pkg/admission/customresourcevalidation/clusterresourcequota/validation/validation.go (hunk 1 above)
Comment (deads2k) [reply pkg/admission/customresourcevalidation/clusterresourcequota/validation/validation.go]: > @derekwaynecarr @deads2k this need to be reviewed as it is not using the ordered map hack we have on internal type.

this was the only significant diff?

Comment (mfojtik) [reply pkg/admission/customresourcevalidation/clusterresourcequota/validation/validation.go]: so far yes.

**Comment by mfojtik** [comment]
@deads2k this was green before conflict :-)

**Comment by mfojtik** [comment]
@deads2k still green!

**Comment by mfojtik** [comment]
/test all

(now that we have this admission plugin enabled ;-)

**Pull request closed without merging**